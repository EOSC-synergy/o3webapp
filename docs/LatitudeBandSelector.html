<!DOCTYPE html>
<html lang="en">

<head>
  
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title> LatitudeBandSelector</title>

  <script src="https://cdn.jsdelivr.net/gh/google/code-prettify@master/loader/run_prettify.js"></script>
  <script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>
  <script src="./build/entry.js"></script>
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->
  <link href="https://fonts.googleapis.com/css?family=Roboto:100,400,700|Inconsolata,700" rel="stylesheet">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.3/css/all.css" integrity="sha384-UHRtZLI+pbxtHCWp1t77Bi1L4ZtiqrqD80Kn4Z8NTSRyMA2Fd33n5dQ8lWUE00s/" crossorigin="anonymous">
  <link type="text/css" rel="stylesheet" href="https://jmblog.github.io/color-themes-for-google-code-prettify/themes/tomorrow-night.min.css">
  <link type="text/css" rel="stylesheet" href="styles/app.min.css">
  <link type="text/css" rel="stylesheet" href="styles/iframe.css">
  <link type="text/css" rel="stylesheet" href="">
  <script async defer src="https://buttons.github.io/buttons.js"></script>

  
</head>



<body class="layout small-header">
    <div id="stickyNavbarOverlay"></div>
    

<div class="top-nav">
    <div class="inner">
        <a id="hamburger" role="button" class="navbar-burger" aria-label="menu" aria-expanded="false">
            <span aria-hidden="true"></span>
            <span aria-hidden="true"></span>
            <span aria-hidden="true"></span>
        </a>
        <div class="logo">
            
             
                <a href="index.html">
                    <h1 class="navbar-item">O3as Webapp documentation</h1>
                </a>
            
        </div>
        <div class="menu">
            
            <div class="navigation">
                <a
                    href="index.html"
                    class="link"
                >
                    Documentation
                </a>
                
                
                
            </div>
        </div>
    </div>
</div>
    <div id="main">
        <div
            class="sidebar "
            id="sidebarNav"
        >
            
            <nav>
                
                    <h2><a href="index.html">Documentation</a></h2><div class="category"><h3>Modules</h3><ul><li><a href="module-API.html">API</a></li><li><a href="module-ArrayOperations.html">ArrayOperations</a></li><li><a href="module-Constants.html">Constants</a></li><li><a href="module-CSV.html">CSV</a></li><li><a href="module-Math.html">Math</a></li><li><a href="module-ModelNameConverter.html">ModelNameConverter</a></li><li><a href="module-OptionsFormatter.html">OptionsFormatter</a></li><li><a href="module-PDF.html">PDF</a></li><li><a href="module-TextSearch.html">TextSearch</a></li></ul><h3>Components</h3><ul><li><a href="AddModelGroupModal.html">AddModelGroupModal</a></li><li><a href="App.html">App</a></li><li><a href="CookieConsentModal.html">CookieConsentModal</a></li><li><a href="CustomLatitudeSelector.html">CustomLatitudeSelector</a></li><li><a href="DiscardChangesModal.html">DiscardChangesModal</a></li><li><a href="downloadGraphAsCSV.html">downloadGraphAsCSV</a></li><li><a href="DownloadModal.html">DownloadModal</a></li><li><a href="EditModelGroupModal.html">EditModelGroupModal</a></li><li><a href="ErrorMessageModal.html">ErrorMessageModal</a></li><li><a href="Footer.html">Footer</a></li><li><a href="Graph.html">Graph</a></li><li><a href="LandingPage.html">LandingPage</a></li><li><a href="LatitudeBandSelector.html">LatitudeBandSelector</a></li><li><a href="Logo.html">Logo</a></li><li><a href="ModelGroupCard.html">ModelGroupCard</a></li><li><a href="ModelGroupConfigurator.html">ModelGroupConfigurator</a></li><li><a href="Navbar.html">Navbar</a></li><li><a href="PlotNameField.html">PlotNameField</a></li><li><a href="PlotTypeSelector.html">PlotTypeSelector</a></li><li><a href="ReferenceModelSelector.html">ReferenceModelSelector</a></li><li><a href="ReferenceYearField.html">ReferenceYearField</a></li><li><a href="RegionSelector.html">RegionSelector</a></li><li><a href="SearchBar.html">SearchBar</a></li><li><a href="SeasonCheckBoxGroup.html">SeasonCheckBoxGroup</a></li><li><a href="Section.html">Section</a></li><li><a href="Sidebar.html">Sidebar</a></li><li><a href="TimeCheckBoxGroup.html">TimeCheckBoxGroup</a></li><li><a href="XAxisField.html">XAxisField</a></li><li><a href="YAxisField.html">YAxisField</a></li></ul><h3>Global</h3><ul><li><a href="global.html#%253C%253C%253C%253C%253C%253C%253C"><<<<<<<</a></li><li><a href="global.html#%253E%253E%253E%253E%253E%253E%253E">>>>>>>></a></li><li><a href="global.html#downloadBase64File">downloadBase64File</a></li><li><a href="global.html#downloadCsvFile">downloadCsvFile</a></li><li><a href="global.html#downloadGraphAsPDF">downloadGraphAsPDF</a></li><li><a href="global.html#downloadGraphAsPNG">downloadGraphAsPNG</a></li><li><a href="global.html#downloadGraphAsSVG">downloadGraphAsSVG</a></li><li><a href="global.html#getAdjustedSVG">getAdjustedSVG</a></li><li><a href="global.html#getLinePresentation">getLinePresentation</a></li><li><a href="global.html#getListOfModelsForPdf">getListOfModelsForPdf</a></li><li><a href="global.html#selectUserRegionName">selectUserRegionName</a></li></ul></div><div class="category"><h2>ModelSlice</h2><h3>Global</h3><ul><li><a href="global.html"></a></li><li><a href="global.html#initialState">initialState</a></li><li><a href="global.html#MODEL_DATA_TEMPLATE">MODEL_DATA_TEMPLATE</a></li><li><a href="global.html#modelsSlice">modelsSlice</a></li><li><a href="global.html#selectAllGroupIds">selectAllGroupIds</a></li><li><a href="global.html#selectAllModelGroups">selectAllModelGroups</a></li><li><a href="global.html#selectModelDataOfGroup">selectModelDataOfGroup</a></li><li><a href="global.html#selectModelsOfGroup">selectModelsOfGroup</a></li><li><a href="global.html#selectNameOfGroup">selectNameOfGroup</a></li><li><a href="global.html#selectStatisticalValueSettingsOfGroup">selectStatisticalValueSettingsOfGroup</a></li><li><a href="global.html#selectVisibilityOfGroup">selectVisibilityOfGroup</a></li></ul></div><div class="category"><h2>plotSlice</h2><h3>Global</h3><ul><li><a href="global.html"></a></li><li><a href="global.html#initialState">initialState</a></li><li><a href="global.html#plotSlice">plotSlice</a></li><li><a href="global.html#selectPlotId">selectPlotId</a></li><li><a href="global.html#selectPlotLocation">selectPlotLocation</a></li><li><a href="global.html#selectPlotMonths">selectPlotMonths</a></li><li><a href="global.html#selectPlotTitle">selectPlotTitle</a></li><li><a href="global.html#selectPlotXRange">selectPlotXRange</a></li><li><a href="global.html#selectPlotYRange">selectPlotYRange</a></li></ul></div><div class="category"><h2>referenceSlice</h2><h3>Global</h3><ul><li><a href="global.html"></a></li><li><a href="global.html#initialState">initialState</a></li><li><a href="global.html#referenceSlice">referenceSlice</a></li><li><a href="global.html#selectIsOffsetApplied">selectIsOffsetApplied</a></li><li><a href="global.html#selectRefModel">selectRefModel</a></li><li><a href="global.html#selectRefYear">selectRefYear</a></li><li><a href="global.html#selectVisibility">selectVisibility</a></li></ul></div>
                
            </nav>
        </div>
        <div class="core" id="main-content-wrapper">
            <div class="content">
                <header class="page-title">
                    <p>Components</p>
                    <h1>LatitudeBandSelector</h1>
                </header>
                




<section>

<header>
    
        <h2>
            &ltLatitudeBandSelector <i>reportError</i> /&gt
        </h2>
        
    
</header>

<article>
    <div class="container-overview">
    
        
            <div class='vertical-section'>
                <div class="members">
                    <div class="member">
                        <div class=name>
                            <span class="tag">Constructor</span>
                        </div>
                        


    
    <h4 class="name" id="LatitudeBandSelector">
        <a class="href-link" href="#LatitudeBandSelector">#</a>
        
        <span class="code-name">
            
                &ltLatitudeBandSelector <i>reportError</i> /&gt
            
        </span>
    </h4>
    

    
    
    <div class="description">
        <p>Enables the user to choose minimum and maximum latitude.
Compromised of <a href="CustomLatitudeSelector.html">CustomLatitudeSelector</a>.</p>
    </div>
    







    
        <h5>PropTypes:</h5>
        
<div class="table-container">
  <table class="params table">
    <thead>
    <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Required</th>
        <th>Description</th>
        <th class="last">Default</th>
    </tr>
    </thead>

    <tbody>
    
          <tr class="deep-level-0">
            <td class="name"><code>reportError</code></td>
            <td>
              func
            </td>
            <td>No</td>
            <td>function for error handling</td>
            <td></td>
          </tr>
          
    
    </tbody>
  </table>
</div>
    



    <h5>Parameters:</h5>
    
<div class="table-container">
    <table class="params table">
        <thead>
        <tr>
            
            <th>Name</th>
            

            <th>Type</th>

            

            

            <th class="last">Description</th>
        </tr>
        </thead>

        <tbody>
        

            
<tr class="deep-level-0">
  
      <td class="name"><code>props</code></td>
  

  <td class="type">
  
      
<code class="param-type">Object</code>


  
  </td>

  

  

  <td class="description last"><p>specified in propTypes</p></td>
</tr>


        
        </tbody>
    </table>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html" class="button">View Source</a>
            <span>
                <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html">views/landingPage/Sidebar/InputComponents/LatitudeBandSelector/LatitudeBandSelector.js</a>, <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html#line30">line 30</a>
            </span>
        </p>
    
</dl>


















    <div class='columns method-parameter'>
        <div class="column is-2"><label>Returns:</label></div>
        <div class="column is-10">
            
                    

<div class="columns">
    
    <div class='param-desc column is-7'><p>a JSX containing a dropdown and if &quot;individual latitude band&quot; is selected a number input field</p></div>
    
    
    <div class='column is-5 has-text-left'>
        <label>Type: </label>
        
<code class="param-type">JSX.Element</code>


    </div>
    
</div>

                
        </div>
    </div>





                    </div>
                </div>
            </div>
        
    
    </div>
    
        
    
    

    

    

    

    

    

    
        <div class='vertical-section'>
            <h1>Members</h1>
            <div class="members">
            
                <div class="member">

    <span class="method-parameter is-pulled-right">
        <label>Type:</label>
        
<code class="param-type">array</code>


    </span>

<h4 class="name" id=".exports.latitudeBands">
    <a class="href-link" href="#.exports.latitudeBands">#</a>
    
        
            <span class='tag'>static</span>
        
            <span class='tag'>constant</span>
        
    
    <span class="code-name">
        exports.latitudeBands
    </span>
    
</h4>




<div class="description">
    <p>Stores the latitude bands and its min/max values</p>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="utils_constants.js.html" class="button">View Source</a>
            <span>
                <a href="utils_constants.js.html">utils/constants.js</a>, <a href="utils_constants.js.html#line157">line 157</a>
            </span>
        </p>
    
</dl>





</div>
            
                <div class="member">

    <span class="method-parameter is-pulled-right">
        <label>Type:</label>
        
<code class="param-type">function</code>


    </span>

<h4 class="name" id=".findLatitudeBandByLocation">
    <a class="href-link" href="#.findLatitudeBandByLocation">#</a>
    
        
            <span class='tag'>static</span>
        
            <span class='tag'>constant</span>
        
    
    <span class="code-name">
        findLatitudeBandByLocation
    </span>
    
</h4>




<div class="description">
    <p>Finds selectedLocation in latitudeBands.</p>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html" class="button">View Source</a>
            <span>
                <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html">views/landingPage/Sidebar/InputComponents/LatitudeBandSelector/LatitudeBandSelector.js</a>, <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html#line107">line 107</a>
            </span>
        </p>
    
</dl>





</div>
            
                <div class="member">

<h4 class="name" id=".selectedLocation">
    <a class="href-link" href="#.selectedLocation">#</a>
    
        
            <span class='tag'>static</span>
        
    
    <span class="code-name">
        selectedLocation
    </span>
    
</h4>




<div class="description">
    <p>An object containing the current minLat and maxLat Values.</p>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html" class="button">View Source</a>
            <span>
                <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html">views/landingPage/Sidebar/InputComponents/LatitudeBandSelector/LatitudeBandSelector.js</a>, <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html#line15">line 15</a>
            </span>
        </p>
    
</dl>





</div>
            
                <div class="member">

<h4 class="name" id="~dispatch">
    <a class="href-link" href="#~dispatch">#</a>
    
        
            <span class='tag'>inner</span>
        
            <span class='tag'>constant</span>
        
    
    <span class="code-name">
        dispatch
    </span>
    
</h4>




<div class="description">
    <p>A dispatch function to dispatch actions to the redux store.</p>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html" class="button">View Source</a>
            <span>
                <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html">views/landingPage/Sidebar/InputComponents/LatitudeBandSelector/LatitudeBandSelector.js</a>, <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html#line39">line 39</a>
            </span>
        </p>
    
</dl>





</div>
            
            </div>
        </div>
    

    
        <div class='vertical-section'>
            <h1>Methods</h1>
            <div class="members">
            
                <div class="member">


    
    <h4 class="name" id="~handleChangeLatitudeBand">
        <a class="href-link" href="#~handleChangeLatitudeBand">#</a>
        
            
                <span class='tag'>inner</span>
            
        
        <span class="code-name">
            
                handleChangeLatitudeBand<span class="signature">(event)</span><span class="type-signature"></span>
            
        </span>
    </h4>
    

    
    
    <div class="description">
        <p>handles the change when the user clicked on a new latitude band option
if the user selected custom sets isCustomizable to true</p>
    </div>
    









    <h5>Parameters:</h5>
    
<div class="table-container">
    <table class="params table">
        <thead>
        <tr>
            
            <th>Name</th>
            

            <th>Type</th>

            

            

            <th class="last">Description</th>
        </tr>
        </thead>

        <tbody>
        

            
<tr class="deep-level-0">
  
      <td class="name"><code>event</code></td>
  

  <td class="type">
  
      
<code class="param-type">event</code>


  
  </td>

  

  

  <td class="description last"><p>the event that triggered this function call</p></td>
</tr>


        
        </tbody>
    </table>
</div>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    
    

    

    

    
        <p class="tag-source">
            <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html" class="button">View Source</a>
            <span>
                <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html">views/landingPage/Sidebar/InputComponents/LatitudeBandSelector/LatitudeBandSelector.js</a>, <a href="views_landingPage_Sidebar_InputComponents_LatitudeBandSelector_LatitudeBandSelector.js.html#line46">line 46</a>
            </span>
        </p>
    
</dl>





















</div>
            
            </div>
        </div>
    

    

    
</article>

</section>




            </div>
            
            <footer class="footer">
                <div class="content has-text-centered">
                    <p>Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.6.10</a></p>
                    <p class="sidebar-created-by">
                        <a href="https://github.com/SoftwareBrothers/better-docs" target="_blank">BetterDocs theme</a> provided with <i class="fas fa-heart"></i> by
                        <a href="http://softwarebrothers.co" target="_blank">SoftwareBrothers - JavaScript Development Agency</a>
                    </p>
                </div>
            </footer>
            
        </div>
        <div id="side-nav" class="side-nav">
        </div>
    </div>
<script src="scripts/app.min.js"></script>
<script>PR.prettyPrint();</script>
<script src="scripts/linenumber.js"> </script>


</body>
</html>