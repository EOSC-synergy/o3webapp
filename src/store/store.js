import { applyMiddleware, combineReducers, configureStore, createStore } from '@reduxjs/toolkit';
import plotReducer from './plotSlice/plotSlice';
import modelsReducer, { setModelsOfModelGroup } from './modelsSlice/modelsSlice';
import referenceReducer from './referenceSlice/referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { HYDRATE, createWrapper } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';

const bindMiddleware = (middleware) => {
    if (process.env.NODE_ENV !== 'production') {
        const { composeWithDevTools } = require('redux-devtools-extension');
        return composeWithDevTools(applyMiddleware(...middleware));
    }
    return applyMiddleware(...middleware);
};

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state, action) => {
    if (action.type === HYDRATE) {
        const nextState = {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return nextState;
    } else {
        return combinedReducer(state, action);
    }
};

let store;

const initStore = () => {
    store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return store;
};

export { store };

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = configureStore({
        reducer: combinedReducer,
    });
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
