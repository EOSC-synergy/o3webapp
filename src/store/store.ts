import { Action, combineReducers, configureStore, ThunkAction } from '@reduxjs/toolkit';
import plotReducer, { GlobalPlotState } from './plotSlice';
import modelsReducer, { GlobalModelState, setModelsOfModelGroup } from './modelsSlice';
import referenceReducer, { GlobalReferenceState } from './referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { createWrapper, HYDRATE } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';
import { useDispatch, useStore } from 'react-redux';

type GlobalState =
    | undefined
    | (GlobalPlotState & GlobalModelState & GlobalReferenceState & { api: any });

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state: GlobalState, action: { type: string; payload: any }) => {
    if (action.type === HYDRATE) {
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
    } else {
        return combinedReducer(state, action);
    }
};

//let store: ReturnType<typeof configureStore<GlobalState, >>;

const initStore = () => {
    //store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return configureStore({
        reducer,
        devTools: process.env.NODE_ENV !== 'production',
        middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
    });
};

export type AppStore = ReturnType<typeof initStore>;
export type AppState = ReturnType<AppStore['getState']>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>;

export type AppDispatch = AppStore['dispatch'];
export const useAppDispatch = () => useDispatch<AppDispatch>(); // Export a hook that can be reused to resolve types

export const useAppStore = () => useStore<AppState>();

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = initStore();
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
