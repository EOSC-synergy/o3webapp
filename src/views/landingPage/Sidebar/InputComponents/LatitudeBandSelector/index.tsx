import React, { FC, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { selectPlotLocation, setLocation } from 'store/plotSlice';
import { Box, Divider, MenuItem, Select, SelectChangeEvent } from '@mui/material';
import Typography from '@mui/material/Typography';
import { Latitude, latitudeBands } from 'utils/constants';
import { fetchPlotDataForCurrentModels } from 'services/API/apiSlice/apiSlice';
import CustomLatitudeSelector from './CustomLatitudeSelector';

/**
 * Enables the user to choose minimum and maximum latitude
 * @component
 * @returns a JSX containing a dropdown and if "individual latitude band" is selected a number input field
 */
const LatitudeBandSelector: FC = () => {
    /**
     * An object containing the current minLat and maxLat Values.
     * @memberof LatitudeBandSelector
     * @type {Object}
     * @see {@link selectPlotLocation}
     */
    const selectedLocation = useSelector(selectPlotLocation);

    /**
     * whether the user selected to enter a custom latitude band
     * @memberof LatitudeBandSelector
     * @type {Array}
     * @default [null, null]
     */
    const [isCustomizable, setIsCustomizable] = useState(false);

    /**
     * A dispatch function to dispatch actions to the redux store.
     */
    const dispatch = useDispatch();

    /**
     * handles the change when the user clicked on a new latitude band option
     * if the user selected custom sets isCustomizable to true
     * @param {event} event the event that triggered this function call
     */
    const handleChangeLatitudeBand = (event: SelectChangeEvent<Latitude>) => {
        if (event.target.value === 'custom') {
            setIsCustomizable(true);
        } else {
            setIsCustomizable(false);
            const latitude = event.target.value as Latitude;
            dispatch(
                setLocation({
                    minLat: latitude.minLat,
                    maxLat: latitude.maxLat,
                })
            );

            // fetch for tco3_zm and tco3_return
            // @ts-expect-error TODO: dispatch typing<
            dispatch(fetchPlotDataForCurrentModels());
        }
    };

    return (
        <>
            <Divider>
                <Typography>LATITUDE BAND</Typography>
            </Divider>
            <Box sx={{ paddingLeft: '8%', paddingRight: '8%', paddingTop: '3%' }}>
                <Select
                    sx={{ width: '100%' }}
                    id="latitudeBandSelector"
                    value={
                        isCustomizable
                            ? latitudeBands[latitudeBands.length - 1].value
                            : findLatitudeBandByLocation(
                                  isCustomizable,
                                  setIsCustomizable,
                                  selectedLocation,
                                  false
                              )
                    }
                    onChange={handleChangeLatitudeBand}
                    defaultValue={findLatitudeBandByLocation(
                        isCustomizable,
                        setIsCustomizable,
                        selectedLocation,
                        false
                    )}
                    inputProps={{ 'data-testid': 'LatitudeBandSelector-select-region' }}
                >
                    {
                        // maps all latitude bands from constants.js to ´MenuItem´s
                        latitudeBands.map((s, idx) => (
                            // @ts-expect-error MenuItem accepts objects, see https://github.com/mui/material-ui/issues/14286
                            <MenuItem key={idx} value={s.value}>
                                {s.text.description}
                            </MenuItem>
                        ))
                    }
                </Select>
                {isCustomizable && <div style={{ height: '10px' }} />}
                {isCustomizable && <CustomLatitudeSelector />}
            </Box>
        </>
    );
};

export default LatitudeBandSelector;

/**
 * Finds selectedLocation in latitudeBands.
 *
 * @param isCustomizable
 * @param setIsCustomizable
 * @param selectedLocation
 * @param {boolean} forceCustomizable if true, acts like isCustomizable is true - if false, does nothing
 * @returns the location
 * @memberof LatitudeBandSelector
 * @function
 */
const findLatitudeBandByLocation = (
    isCustomizable: boolean,
    setIsCustomizable: (customizable: boolean) => void,
    selectedLocation: Latitude,
    forceCustomizable?: boolean
): Latitude => {
    if (!forceCustomizable) {
        for (let i = 0; i < latitudeBands.length - 1; i++) {
            if (
                latitudeBands[i].value.minLat === selectedLocation.minLat &&
                latitudeBands[i].value.maxLat === selectedLocation.maxLat
            ) {
                return latitudeBands[i].value;
            }
        }
    }
    if (isCustomizable || forceCustomizable) {
        return latitudeBands[latitudeBands.length - 1].value;
    }
    setIsCustomizable(true);
    return findLatitudeBandByLocation(isCustomizable, setIsCustomizable, selectedLocation, true);
};
