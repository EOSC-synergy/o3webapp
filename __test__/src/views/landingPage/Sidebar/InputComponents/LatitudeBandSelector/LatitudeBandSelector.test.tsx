import { render, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import LatitudeBandSelector from 'views/landingPage/Sidebar/InputComponents/LatitudeBandSelector';
import { AppStore, createTestStore } from 'store';

let store: AppStore;
beforeEach(() => {
    store = createTestStore();
});

describe('tests basic rendering', () => {
    it('renders without crashing', () => {
        render(
            <>
                <Provider store={store}>
                    <LatitudeBandSelector />
                </Provider>
            </>
        );
    });

    it('renders correctly', () => {
        const { container } = render(
            <>
                <Provider store={store}>
                    <LatitudeBandSelector />
                </Provider>
            </>
        );

        expect(container).toMatchSnapshot();
    });
});

describe('tests redux functionality', () => {
    it('updates the region accordingly in the store', () => {
        const { getByRole, getAllByRole } = render(
            <Provider store={store}>
                <LatitudeBandSelector />
            </Provider>
        );

        act(() => {
            userEvent.click(getByRole('button'));
        });
        const options = getAllByRole('option');
        act(() => {
            userEvent.click(options[2]);
        });

        expect(store.getState().plot.generalSettings.location).toEqual({ minLat: -20, maxLat: 20 }); // Tropics
        act(() => {
            userEvent.click(options[options.length - 1]);
        });
    });

    it('displays a custom latitude band selector if custom region is selected', () => {
        const { getByRole, getAllByRole, container } = render(
            <Provider store={store}>
                <LatitudeBandSelector />
            </Provider>
        );

        act(() => {
            userEvent.click(getByRole('button'));
        });
        const options = getAllByRole('option');

        expect(container).not.toHaveTextContent('Min. Lat');
        expect(container).not.toHaveTextContent('Max. Lat');
        act(() => {
            userEvent.click(options[options.length - 1]);
        });
        expect(container).toHaveTextContent('Min. Lat');
        expect(container).toHaveTextContent('Max. Lat');
    });
});
