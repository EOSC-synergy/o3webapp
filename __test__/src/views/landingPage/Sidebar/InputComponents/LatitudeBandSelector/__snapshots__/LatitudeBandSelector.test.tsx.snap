// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests basic rendering renders correctly 1`] = `
<div>
  <div
    class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1ws8119-MuiDivider-root"
    role="separator"
  >
    <span
      class="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        LATITUDE BAND
      </p>
    </span>
  </div>
  <div
    class="MuiBox-root css-1n3sbe4"
  >
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary  css-bx5laj-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="latitudeBandSelector"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="latitudeBandSelector"
        role="button"
        tabindex="0"
      >
        Global (90°S–90°N)
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        data-testid="LatitudeBandSelector-select-region"
        tabindex="-1"
        value="[object Object]"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-hdw1oc"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
